qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -160,83 +180,198 @@ def __init__(
         hls_names = set(self.hls_plugin_manager.plugins_by_op)
-        if not hls_names.intersection(dag.count_ops()):
+        node_names = dag.count_ops()
+        if "annotated" not in node_names and not hls_names.intersection(node_names):
             return dag
-        for node in dag.op_nodes():
-            if node.name in self.hls_config.methods.keys():
-                # the operation's name appears in the user-provided config,
-                # we use the list of methods provided by the user
-                methods = self.hls_config.methods[node.name]
-            elif (
-                self.hls_config.use_default_on_unspecified
-                and "default" in self.hls_plugin_manager.method_names(node.name)
-            ):
-                # the operation's name does not appear in the user-specified config,
-                # we use the "default" method when instructed to do so and the "default"
-                # method is available
-                methods = ["default"]
+
+        # copy dag_op_nodes because we are modifying the DAG below
+        dag_op_nodes = dag.op_nodes()
+
+        for node in dag_op_nodes:
+            qubits = (
+                [dag.find_bit(x).index for x in node.qargs] if self._use_qubit_indices else None
+            )
+            decomposition = self._recursively_handle_op(node.op, qubits)
+
+            if not isinstance(decomposition, (QuantumCircuit, Operation)):
+                raise TranspilerError(f"HighLevelSynthesis was unable to synthesize {node.op}.")
+
+            if isinstance(decomposition, QuantumCircuit):
+                dag.substitute_node_with_dag(node, circuit_to_dag(decomposition))
+            elif isinstance(decomposition, Operation):
+                dag.substitute_node(node, decomposition)
+
+        return dag
+
+    def _recursively_handle_op(
+        self, op: Operation, qubits: Optional[List] = None
+    ) -> Union[Operation, QuantumCircuit]:
+        """Recursively synthesizes a single operation.
+
+        The result can be either another operation or a quantum circuit.
+
+        Some examples when the result can be another operation:
+        Adding control to CX-gate results in CCX-gate,
+        Adding inverse to SGate results in SdgGate.
+
+        Some examples when the result can be a quantum circuit:
+        Synthesizing a LinearFunction produces a quantum circuit consisting of
+        CX-gates.
+
+        The function recursively handles operation's definition, if it exists.
+        """
+
+        # First, try to apply plugin mechanism
+        decomposition = self._synthesize_op_using_plugins(op, qubits)
+        if decomposition:
+            return decomposition
+
+        # Second, handle annotated operations
+        # For now ignore the qubits over which the annotated operation is defined.
+        decomposition = self._synthesize_annotated_op(op)
+        if decomposition:
+            return decomposition
```

Comment:
Is the reason to use an if/else between plugin and annotated because you might have a plugin for annotated ops?

Reply:
You mean the reason _not_ to use an if/else, right? But yes, at some point @jakelishman has suggested the possibility of supporting annotated ops via a plugin. This would actually make the code a bit cleaner, yet seems to be significantly more challenging to implement while supporting full recursion. Anyhow, I think that it's good to leave this possibility here.